### PR TITLE
Support PEM encoded certificates and EC for private keys

### PR DESCRIPTION
**Pull request by abremora**
Support for PEM encoded certificates added. Although there is a parser for PEM files already available (see `PemObject`), I introduced a new one called `PemReader`. This implementation has no dependencies to RSA. This seems to be reasonable because there are more algorithms available. Further I looked for a solution to add easily more private key implementations. The result was a `PrivateKeyFactory` which holds a list of private key strategies (`RsaPrivateKeyStrategy`, `EcPrivateKeyStrategy`).
`RsaPrivateKeyStrategy` is just a wrapper for `KeystoreUtil.getRSAPrivateKeySpec()`. 
To prove the concept, I quickly develop a `EcPrivateKeyStrategy`. That's why the support for EC is already present here. If you like to split these two features into separate PRs just give me a note.

Fixes gh-678
Fixes gh-683

### CODE REVIEW

File: spring-vault-core/src/main/java/org/springframework/vault/support/Certificate.java [hunk 1]
```diff
@@ -39,103 +38,121 @@
...
+
+  private final CertificateEncoding issuingCaCertificateFormat;
+
+  Certificate(@JsonProperty("serial_number") String serialNumber, @JsonProperty("certificate") String certificate,
+      @JsonProperty("issuing_ca") String issuingCaCertificate) {
+
+    this.serialNumber = serialNumber;
+    this.certificate = certificate;
+    this.issuingCaCertificate = issuingCaCertificate;
+    this.certificateFormat = CertificateEncodingValidator.getFormat(certificate);
```
Comment (abremora) [inline spring-vault-core/src/main/java/org/springframework/vault/support/Certificate.java:57]: We have no information about the current encoding so we have to calculate it for further processing. Should this an enum or a string only?

File: spring-vault-core/src/main/java/org/springframework/vault/support/EcPrivateKeyStrategy.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+ */
+package org.springframework.vault.support;
+
+import java.security.spec.ECFieldFp;
+import java.security.spec.ECParameterSpec;
+import java.security.spec.ECPoint;
+import java.security.spec.ECPrivateKeySpec;
+import java.security.spec.EllipticCurve;
+import java.security.spec.KeySpec;
+import org.bouncycastle.asn1.ASN1ObjectIdentifier;
```
Comment (abremora) [inline spring-vault-core/src/main/java/org/springframework/vault/support/EcPrivateKeyStrategy.java:24]: We use Bouncy Castle here for EC support. Can we use this dependency here?

Comment (mp911de) [reply spring-vault-core/src/main/java/org/springframework/vault/support/EcPrivateKeyStrategy.java:24]: Ideally, we don't use BouncyCastle at all as the cost for the dependency is quite large in comparison to what we get. We have a DER parser already and it's a matter of parsing the ASN.1 sequence. Java's crypto utils contain proper support to resolve EC parameters through an OID and provide key specs.

I have the code ready for this.

**Comment by abremora** [comment]
I am happy to support you. Please do not hesitate to contact me if you need further assistance.

**Pull request closed without merging**